crates/tlsn/src/prover.rs

```diff
@@ -202,95 +198,135 @@ impl Prover<state::CommitAccepted> {
+        self.state.tls_client.write_tls(buf)
+    }
 
-                {
-                    let mut vm = vm.try_lock().expect("VM should not be locked");
+    /// Returns `true` if the prover can read plaintext data.
+    pub fn wants_read(&self) -> bool {
+        self.state.tls_client.wants_read()
+    }
 
-                    debug!("finalizing mpc");
+    /// Returns `true` if the prover wants to write plaintext data.
+    pub fn wants_write(&self) -> bool {
+        self.state.tls_client.wants_write()
+    }
 
-                    // Finalize DEAP.
-                    mux_fut
-                        .poll_with(vm.finalize(&mut ctx))
-                        .await
-                        .map_err(ProverError::mpc)?;
+    /// Reads plaintext data from the server into the provided buffer.
+    pub fn read(&mut self, buf: &mut [u8]) -> Result<usize, std::io::Error> {
+        self.state.tls_client.read(buf)
+    }
 
-                    debug!("mpc finalized");
-                }
+    /// Writes plaintext data to be sent to the server.
+    pub fn write(&mut self, buf: &[u8]) -> Result<usize, std::io::Error> {
+        self.state.tls_client.write(buf)
+    }
 
-                // Pull out ZK VM.
-                let (_, mut vm) = Arc::into_inner(vm)
-                    .expect("vm should have only 1 reference")
-                    .into_inner()
-                    .into_inner();
+    /// Closes the connection from the client side.
+    pub fn client_close(&mut self) -> Result<(), std::io::Error> {
+        self.state.tls_client.client_close()
+    }
 
-                // Prove tag verification of received records.
-                // The prover drops the proof output.
-                let _ = verify_tags(
-                    &mut vm,
-                    (keys.server_write_key, keys.server_write_iv),
-                    keys.server_write_mac_key,
-                    *tls_transcript.version(),
-                    tls_transcript.recv().to_vec(),
-                )
-                .map_err(ProverError::zk)?;
+    /// Closes the connection from the server side.
+    pub fn server_close(&mut self) -> Result<(), std::io::Error> {
+        self.state.tls_client.server_close()
+    }
 
-                mux_fut
-                    .poll_with(vm.execute_all(&mut ctx).map_err(ProverError::zk))
-                    .await?;
+    /// Defers decryption of data from the server until the server has closed
+    /// the connection.
+    ///
+    /// This is a performance optimization which will significantly reduce the
+    /// amount of upload bandwidth used by the prover.
+    ///
+    /// # Notes
+    ///
+    /// * The prover may need to close the connection to the server in order for
+    ///   it to close the connection on its end. If neither the prover or server
+    ///   close the connection this will cause a deadlock.
+    pub async fn defer_decryption(&mut self) -> Result<(), ProverError> {
```

Comment:
Let's change this to `enable_decryption(enable: false)`, and add an `is_decrypting` method.

Also, this method should not be async. So we'll need to move the handling of this into `poll`